src/DictList.elm

```diff
@@ -938,6 +965,174 @@ fromDict dict =
+-}
+dropWhile : ((comparable, a) -> Bool) -> DictList comparable a -> DictList comparable a
+dropWhile predicate list =
+  toList list
+    |> List.Extra.dropWhile predicate
+    |> fromList
+
+{-| Remove duplicate values, keeping the first instance of each element which appears more than once.
+
+    unique [0,1,1,0,1] == [0,1]
+-}
+unique : DictList comparable1 comparable2 -> DictList comparable1 comparable2
+unique list =
+  toList list
+    |> List.Extra.unique
+    |> fromList
+
+{-| Drop duplicates where what is considered to be a duplicate is the result of first applying the supplied function to the elements of the list.
+-}
+uniqueBy : (comparable1 -> a -> comparable2) -> DictList comparable1 a -> DictList comparable1 a
+uniqueBy f list =
+  toList list
+    |> List.Extra.uniqueBy (uncurry f)
+    |> fromList
+
+{-| Indicate if list has duplicate values.
+
+    allDifferent [0,1,1,0,1] == False
+-}
+allDifferent : DictList comparable1 comparable2 -> Bool
+allDifferent list =
+  -- @TODO Should this method just use the values, or also the keys
+  values list
+   |> List.Extra.allDifferent 
+
+{-| Indicate if list has duplicate values when supplied function are applyed on each values.
+-}
+allDifferentBy : (comparable1 -> a -> comparable2) -> DictList comparable1 a -> Bool
+allDifferentBy f list = 
+  toList list
+    |> List.Extra.allDifferentBy (uncurry f)
+
+{-| Map functions taking multiple arguments over multiple lists. Each list should be of the same length.
+
+    ((\a b c -> a + b * c)
+        |> flip map [1,2,3]
+        |> andMap [4,5,6]
+        |> andMap [2,1,1]
+    ) == [9,7,9]
+-}
+andMap : DictList comparable a -> DictList comparable (a -> b) -> DictList comparable b
+andMap l fl =
+  let
+    keyList = keys l
+    lList = values l
+    flList = values fl
+  in
+      List.Extra.andMap lList flList
+      |> List.Extra.zip keyList
+      |> fromList
+
+{-| Equivalent to `concatMap`. For example, suppose you want to have a cartesian product of [1,2] and [3,4]:
+
+    [1,2] |> andThen (\x -> [3,4]
+          |> andThen (\y -> [(x,y)]))
+
+will give back the list:
+
+    [(1,3),(1,4),(2,3),(2,4)]
+
+Now suppose we want to have a cartesian product between the first list and the second list and its doubles:
+
+    [1,2] |> andThen (\x -> [3,4]
+          |> andThen (\y -> [y,y*2]
+          |> andThen (\z -> [(x,z)])))
+
+will give back the list:
+
+    [(1,3),(1,6),(1,4),(1,8),(2,3),(2,6),(2,4),(2,8)]
+
+Advanced functional programmers will recognize this as the implementation of bind operator (>>=) for lists from the `Monad` typeclass.
+-}
+andThen : (comparable -> a -> DictList comparable b) -> DictList comparable a -> DictList comparable b
+andThen =
+    concatMap
+
+concatMap : (comparable -> a -> DictList comparable b) -> DictList comparable a -> DictList comparable b
+concatMap f xs = empty
+--  map f xs
+--    |> toList
+--    |> values
+--    |> concat
+
```

Comment:
This is such an interesting question! I've been thinking about that and talking with a bunch of people.

One sensefull type would look like the following: ```(DictList comparable (a -> b)) -> (DictList comparable a) -> (DictList comparable b)```. For the concrete implementation you would probably match the keys and throw away all (a->b)'s / a's when there are no matching keys. This requires for example no instance of ```pure``` (the other function which makes sense in the context). To implement ```pure``` you would need some ```mempty``` for the keys.

Note: Dict.Map doesn't implement that in haskell
Note2: https://hackage.haskell.org/package/total-map-0.0.6/docs/Data-TotalMap.html does implement the above idea

For ```andThen``` the signature could look like the following:

```andThen : (comparable -> a -> DictList comparable b) -> DictList comparable a -> DictList comparable b```

What could this function do:

* This function maps a key/value pair to a new DictList, so it kinda generates a new dict list.
* The result could then have the specific key removed and inserted all keys/values of the result in b. This smashing together could be some kind of ```merge``` operation.

In this world you could reimplement:

* ```remove k dictList = andThen (\k2 v -> if k == k2 then empty else fromList [(k2, v)]) dictList```
* ```update k v dictList = andThen (\k2 v2 -> fromList [(k2, if k == k2 then v else v2)]) dictList```

Does it even remotely makes sense?

__TL;DR__

* Don't expect to have some form of product/cartesian behaviour for ```andMap```
* I doubt there is too much value in ```return``` / ```pure``` implementations